core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -51,6 +49,9 @@ private[spark] abstract class YarnSchedulerBackend(
 
   private implicit val askTimeout = RpcUtils.askRpcTimeout(sc.conf)
 
+  // Flag to specify whether AM is registered or not.
+  @volatile private var isAmRegistered = false
```

Comment:
You don't need this. If the AM hasn't registered before, `amEndpoint` will be `None`. Just check that.